### PR TITLE
Expiring multi party `siphon`

### LINKED ISSUE #1001: Add "siphon" function from leftover collateral in a contract post expiry
If collateral has been left in an expiring multi party for a long period of time and has not been claimed there should be a method to withdraw this.

### PR DESCRIPTION
**Pull request by chrismaree**
This PR introduces the notion of a `siphon` function that enables excess collateral to be drained out of a contract if it has not been redeemed after a set maturity. All drained funds are sent to the store. 

The implementation in this PR was informed from a discussion we had during the week. If we choose to change how this siphon function works it's not a huge deal as this PR lays the foundation for this process. We can always introduce more logic to it as well as additional governance steps if we deem it appropriate.

Post `siphonDelay` all functionality is disabled except for the `siphon` function as the `onlyPostExpiration` modifier operates within the range from:

```
expiatory < t < expiry + siphonDelay
```

close #1001

### CODE REVIEW

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 1]
```diff
@@ -58,6 +58,9 @@ contract PricelessPositionManager is FeePayer {
     uint public expirationTimestamp;
     // Time that has to elapse for a withdrawal request to be considered passed, if no liquidations occur.
     uint public withdrawalLiveness;
+    // Time that has to elapse for the contract to be siphoned. This occures if a long period (>6 months) has
+    // occured post settlement and there is remaining unclaimed collateral in the contract.
```
Comment (nicholaspai) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: nit: should be post "expiration" not post "settlement" right? Just want to make sure I'm understanding properly

Comment (chrismaree) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: in this context is settlement not the same expiration?

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 2]
```diff
@@ -366,6 +371,17 @@ contract PricelessPositionManager is FeePayer {
...
     }
 
+    /**
+     * @notice If token holders or sponsors do not redeem their positions for underlying collateral within
+     * a pre-defined `siphonDelay`, all underlying collateral is sent to the Store. This delay is set to
+     * be sufficiently long such that token holders have a reasonable period of time to withdrawal (>6 months).
+     */
+    function siphonContractCollateral() external {
+        require(getCurrentTime() > expirationTimestamp + siphonDelay);
+        uint contractCollateralBallance = collateralCurrency.balanceOf(address(this));
```
Comment (nicholaspai) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: typo `Balance`

**Comment by chrismaree** [comment]
> Requesting a second eye from @ptare--
> 
> Do fees continue to accrue between expirationTimestamp and siphonTimestamp?

I think they do, yes

**Comment by nicholaspai** [comment]
> > Requesting a second eye from @ptare--
> > Do fees continue to accrue between expirationTimestamp and siphonTimestamp?
> 
> I think they do, yes

OK, I _think_ this doesn't raise any issues

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 3]
```diff
@@ -366,6 +371,17 @@ contract PricelessPositionManager is FeePayer {
...
 
+    /**
+     * @notice If token holders or sponsors do not redeem their positions for underlying collateral within
+     * a pre-defined `siphonDelay`, all underlying collateral is sent to the Store. This delay is set to
+     * be sufficiently long such that token holders have a reasonable period of time to withdrawal (>6 months).
+     */
+    function siphonContractCollateral() external {
+        require(getCurrentTime() > expirationTimestamp + siphonDelay);
+        uint contractCollateralBalance = collateralCurrency.balanceOf(address(this));
+        collateralCurrency.safeTransfer(_getStoreAddress(), contractCollateralBalance);
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: Can you write up (maybe in the PR description) what happens to all the contract functions after this is called? Are there any unintended consequences wrt contract bookkeeping or user interaction?

Comment (chrismaree) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: done

**Comment by nicholaspai** [comment]
I believe this partially addresses #925 , but in reviewing that issue, I'd ask again whether we account for case where whether through regular fees or other means, a contract needs to `requestOraclePrice` but doesn't have enough collateral. @mrice32 @chrismaree thoughts?

**Comment by chrismaree** [comment]
> I believe this partially addresses #925 , but in reviewing that issue, I'd ask again whether we account for case where whether through regular fees or other means, a contract needs to `requestOraclePrice` but doesn't have enough collateral. @mrice32 @chrismaree thoughts?

It is partially addressed I think. We should separately look at that problem though.

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 4]
```diff
@@ -379,6 +384,17 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
         emit SettleExpiredPosition(msg.sender, totalRedeemableCollateral.rawValue, tokensToRedeem.rawValue);
     }
 
+    /**
+     * @notice If token holders or sponsors do not redeem their positions for underlying collateral within
+     * a pre-defined `siphonDelay`, all underlying collateral is sent to the Store. This delay is set to
+     * be sufficiently long such that token holders have a reasonable period of time to withdrawal (>6 months).
+     */
+    function siphonContractCollateral() external {
+        require(getCurrentTime() > expirationTimestamp + siphonDelay);
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]:
```suggestion
        require(getCurrentTime() > expirationTimestamp.add(siphonDelay));
```

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 5]
```diff
@@ -525,6 +546,10 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
         require(getCurrentTime() >= expirationTimestamp);
     }
 
+    function _onlyPreSiphon() internal view {
+        require(getCurrentTime() < expirationTimestamp + siphonDelay);
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]:
```suggestion
        require(getCurrentTime() < expirationTimestamp.add(siphonDelay));
```

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 6]
```diff
@@ -379,6 +384,17 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
         emit SettleExpiredPosition(msg.sender, totalRedeemableCollateral.rawValue, tokensToRedeem.rawValue);
     }
 
+    /**
+     * @notice If token holders or sponsors do not redeem their positions for underlying collateral within
+     * a pre-defined `siphonDelay`, all underlying collateral is sent to the Store. This delay is set to
+     * be sufficiently long such that token holders have a reasonable period of time to withdrawal (>6 months).
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: Small nit: we don't technically control how large this value is -- the deployer does.
```suggestion
     * a pre-defined `siphonDelay`, all underlying collateral is sent to the Store. This delay should be set to
     * be sufficiently long such that token holders have a reasonable period of time to withdrawal (>6 months).
```

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 7]
```diff
@@ -59,6 +59,9 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
     uint public expirationTimestamp;
     // Time that has to elapse for a withdrawal request to be considered passed, if no liquidations occur.
     uint public withdrawalLiveness;
+    // Time that has to elapse for the contract to be siphoned. This occures if a long period (>6 months) has
+    // occured post expiration and there is remaining unclaimed collateral in the contract.
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: I don't think we can (or should) assume the siphonDelay is long or > 6 months.

Comment (chrismaree) [reply core/contracts/financial-templates/implementation/PricelessPositionManager.sol]: have removed this part of the comment

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 8]
```diff
@@ -379,6 +384,17 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
         emit SettleExpiredPosition(msg.sender, totalRedeemableCollateral.rawValue, tokensToRedeem.rawValue);
     }
 
+    /**
+     * @notice If token holders or sponsors do not redeem their positions for underlying collateral within
+     * a pre-defined `siphonDelay`, all underlying collateral is sent to the Store. This delay is set to
+     * be sufficiently long such that token holders have a reasonable period of time to withdrawal.
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]:
```suggestion
     * a pre-defined `siphonDelay`, all underlying collateral is sent to the Store. This delay
     * should be set to be sufficiently long such that token holders have a reasonable
     * period of time to withdrawal.
```

File: core/contracts/financial-templates/implementation/PricelessPositionManager.sol [hunk 9]
```diff
@@ -59,6 +59,9 @@ contract PricelessPositionManager is FeePayer, AdministrateeInterface {
     uint public expirationTimestamp;
     // Time that has to elapse for a withdrawal request to be considered passed, if no liquidations occur.
     uint public withdrawalLiveness;
+    // Time that has to elapse for the contract to be siphoned. This occures if a long period has
+    // occured post expiration and there is remaining unclaimed collateral in the contract.
```
Comment (mrice32) [inline core/contracts/financial-templates/implementation/PricelessPositionManager.sol]:
```suggestion
    // Time that has to elapse for the contract to be siphoned. This occurs if `siphonDelay`
    // seconds have past after the expiration timestamp and there is remaining unclaimed
    // collateral in the contract.
```

**Comment by chrismaree** [comment]
> Still wanted to tweak those comments a little. I don't think we want to assume anything about the siphonDelay that's passed in since the deployer could set it anywhere between 0 and inf.

Ok cool. I have accepted

**Pull request merged**